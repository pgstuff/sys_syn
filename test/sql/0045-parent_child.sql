BEGIN;

CREATE EXTENSION tinyint
    SCHEMA public;

CREATE EXTENSION sys_syn;

CREATE SCHEMA user_data
    AUTHORIZATION postgres;

CREATE TABLE user_data.parent_table (
        parent_table_key integer NOT NULL,
        parent_table_text text,
        CONSTRAINT parent_table_pkey PRIMARY KEY (parent_table_key));

CREATE TABLE user_data.child_table (
        child_table_key integer NOT NULL,
        parent_table_key integer,
        CONSTRAINT child_table_pkey PRIMARY KEY (child_table_key));

INSERT INTO sys_syn.in_groups_def VALUES ('in');

SELECT sys_syn.in_table_add_sql('user_data.parent_table'::regclass, 'in');

DO $$BEGIN
        EXECUTE sys_syn.in_table_add_sql('user_data.parent_table'::regclass, 'in');
END$$;

SELECT sys_syn.in_table_add (
                schema          => 'user_data',
                in_table_id     => 'child_table',
                in_group_id     => 'in',
                in_pull_id      => NULL,
                in_columns      => ARRAY[
                       $COL$("child_table_key","integer",Key,"in_source.child_table_key",,,,)$COL$,
                       $COL$("parent_table_key","integer",Attribute,"in_source.parent_table_key",,1,"parent_table","parent_table_key")$COL$
                ]::sys_syn.create_in_column[],
                full_table_reference    => 'user_data.child_table'
        );

INSERT INTO user_data.parent_table(
        parent_table_key,       parent_table_text)
VALUES (1,                      'parent_data');

INSERT INTO user_data.child_table(
        child_table_key,        parent_table_key)
VALUES (2,                      1);

INSERT INTO sys_syn.out_groups_def VALUES ('out');

SELECT sys_syn.out_table_add('user_data', 'parent_table', 'out');

SELECT sys_syn.out_table_add('user_data', 'child_table', 'out');

ALTER TABLE user_data.parent_table_out_queue
  ADD FOREIGN KEY (trans_id_in, key) REFERENCES user_data.parent_table_in (trans_id_in, key) ON UPDATE RESTRICT ON DELETE RESTRICT;

SELECT user_data.parent_table_pull(FALSE);
SELECT user_data.parent_table_out_move();
SELECT key, delta_type, queue_state FROM user_data.parent_table_out_queue;

UPDATE sys_syn.trans_id_mod SET trans_id_mod = trans_id_mod + 1;
SELECT user_data.child_table_pull(FALSE);
SELECT user_data.child_table_out_move();
SELECT key, delta_type, queue_state FROM user_data.child_table_out_queue;

UPDATE user_data.parent_table_out_queue SET queue_state = 'Reading'::sys_syn.queue_state WHERE (key).parent_table_key = 1;
UPDATE user_data.parent_table_out_queue SET queue_state = 'Processed'::sys_syn.queue_state WHERE (key).parent_table_key = 1;
SELECT user_data.parent_table_out_processed();
SELECT key, delta_type, queue_state FROM user_data.parent_table_out_queue;

UPDATE sys_syn.trans_id_mod SET trans_id_mod = trans_id_mod + 1;
SELECT user_data.child_table_pull(FALSE);
SELECT user_data.child_table_out_move();
SELECT key, delta_type, queue_state FROM user_data.child_table_out_queue;

UPDATE user_data.child_table_out_queue SET queue_state = 'Reading'::sys_syn.queue_state WHERE (key).child_table_key = 2;
UPDATE user_data.child_table_out_queue SET queue_state = 'Processed'::sys_syn.queue_state WHERE (key).child_table_key = 2;
SELECT user_data.child_table_out_processed();
SELECT key, delta_type, queue_state FROM user_data.parent_table_out_queue;

ROLLBACK;
