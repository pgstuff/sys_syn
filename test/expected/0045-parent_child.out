BEGIN;
CREATE EXTENSION tinyint SCHEMA public;
CREATE EXTENSION sys_syn;
CREATE SCHEMA user_data
    AUTHORIZATION postgres;
CREATE TABLE user_data.parent_table (
        parent_table_id integer NOT NULL,
        parent_table_text text,
        CONSTRAINT parent_table_pid PRIMARY KEY (parent_table_id));
CREATE TABLE user_data.child_table (
        child_table_id integer NOT NULL,
        parent_table_id integer,
        CONSTRAINT child_table_pid PRIMARY KEY (child_table_id));
INSERT INTO sys_syn.in_groups_def VALUES ('in');
SELECT sys_syn.in_table_create_sql('user_data.parent_table'::regclass, 'in');
                                            in_table_create_sql                                             
------------------------------------------------------------------------------------------------------------
 SELECT  sys_syn.in_table_create(                                                                          +
                 schema          => 'user_data'::regnamespace,                                             +
                 in_table_id     => 'parent_table',                                                        +
                 in_group_id     => 'in',                                                                  +
                 in_pull_id      => NULL,                                                                  +
                 in_columns      => ARRAY[                                                                 +
                        $COL$("parent_table_id","integer",Id,"in_source.parent_table_id",,,,,)$COL$,       +
                        $COL$("parent_table_text","text",Attribute,"in_source.parent_table_text",,,,,)$COL$+
                 ]::sys_syn.create_in_column[],                                                            +
                 full_table_reference    => 'user_data.parent_table',                                      +
                 changes_table_reference => NULL,                                                          +
                 full_sql                => NULL,                                                          +
                 changes_sql             => NULL,                                                          +
                 full_pre_sql            => NULL,                                                          +
                 changes_pre_sql         => NULL,                                                          +
                 full_post_sql           => NULL,                                                          +
                 changes_post_sql        => NULL,                                                          +
                 enable_deletes_implied  => 'true',                                                        +
                 null_key_handler        => 'none'::sys_syn.null_key_handler,                              +
                 key_violation_handler   => 'none'::sys_syn.key_violation_handler,                         +
                 full_prepull_id         => NULL,                                                          +
                 changes_prepull_id      => NULL,                                                          +
                 record_comparison_different=>NULL,                                                        +
                 record_comparison_same  => NULL,                                                          +
                 tablespace              => NULL,                                                          +
                 in_partitions           => ARRAY[                                                         +
                         $PART$("",)$PART$]::sys_syn.create_in_partition[]                                 +
         );
(1 row)

DO $$BEGIN
        EXECUTE sys_syn.in_table_create_sql('user_data.parent_table'::regclass, 'in');
END$$;
SELECT sys_syn.in_table_create (
                schema          => 'user_data',
                in_table_id     => 'child_table',
                in_group_id     => 'in',
                in_pull_id      => NULL,
                in_columns      => ARRAY[
                       $COL$("child_table_id","integer",Id,"in_source.child_table_id",,,,,)$COL$,
                       $COL$("parent_table_id","integer",Attribute,"in_source.parent_table_id",,1,"parent_table","parent_table_id",)$COL$
                ]::sys_syn.create_in_column[],
                full_table_reference    => 'user_data.child_table'
        );
 in_table_create 
-----------------
 
(1 row)

INSERT INTO user_data.parent_table(
        parent_table_id,       parent_table_text)
VALUES (1,                      'parent_data');
INSERT INTO user_data.child_table(
        child_table_id,        parent_table_id)
VALUES (2,                      1);
INSERT INTO sys_syn.out_groups_def VALUES ('out');
SELECT sys_syn.out_table_create('user_data', 'parent_table', 'out');
 out_table_create 
------------------
 
(1 row)

SELECT sys_syn.out_table_create('user_data', 'child_table', 'out');
 out_table_create 
------------------
 
(1 row)

ALTER TABLE user_data.parent_table_out_queue_1
  ADD FOREIGN KEY (trans_id_in, id) REFERENCES user_data.parent_table_in_1 (trans_id_in, id) ON UPDATE RESTRICT ON DELETE RESTRICT;
SELECT user_data.parent_table_pull(FALSE);
 parent_table_pull 
-------------------
 t
(1 row)

SELECT user_data.parent_table_out_move_1();
 parent_table_out_move_1 
-------------------------
 t
(1 row)

SELECT id, delta_type, queue_state FROM user_data.parent_table_out_queue_1;
 id  | delta_type | queue_state 
-----+------------+-------------
 (1) | Add        | Unclaimed
(1 row)

UPDATE sys_syn.trans_id_mod SET trans_id_mod = trans_id_mod + 1;SET LOCAL sys_syn.trans_id_curr TO 2;
SELECT user_data.child_table_pull(FALSE);
 child_table_pull 
------------------
 t
(1 row)

SELECT user_data.child_table_out_move_1();
 child_table_out_move_1 
------------------------
 f
(1 row)

SELECT id, delta_type, queue_state FROM user_data.child_table_out_queue_1;
 id | delta_type | queue_state 
----+------------+-------------
(0 rows)

UPDATE user_data.parent_table_out_queue_1 SET queue_state = 'Claimed'::sys_syn.queue_state WHERE (id).parent_table_id = 1;
UPDATE user_data.parent_table_out_queue_1 SET queue_state = 'Processed'::sys_syn.queue_state WHERE (id).parent_table_id = 1;
SELECT user_data.parent_table_out_processed_1();
 parent_table_out_processed_1 
------------------------------
 t
(1 row)

SELECT id, delta_type, queue_state FROM user_data.parent_table_out_queue_1;
 id | delta_type | queue_state 
----+------------+-------------
(0 rows)

UPDATE sys_syn.trans_id_mod SET trans_id_mod = trans_id_mod + 1;SET LOCAL sys_syn.trans_id_curr TO 3;
SELECT user_data.child_table_pull(FALSE);
 child_table_pull 
------------------
 t
(1 row)

SELECT user_data.child_table_out_move_1();
 child_table_out_move_1 
------------------------
 t
(1 row)

SELECT id, delta_type, queue_state FROM user_data.child_table_out_queue_1;
 id  | delta_type | queue_state 
-----+------------+-------------
 (2) | Add        | Unclaimed
(1 row)

UPDATE user_data.child_table_out_queue_1 SET queue_state = 'Claimed'::sys_syn.queue_state WHERE (id).child_table_id = 2;
UPDATE user_data.child_table_out_queue_1 SET queue_state = 'Processed'::sys_syn.queue_state WHERE (id).child_table_id = 2;
SELECT user_data.child_table_out_processed_1();
 child_table_out_processed_1 
-----------------------------
 t
(1 row)

SELECT id, delta_type, queue_state FROM user_data.parent_table_out_queue_1;
 id | delta_type | queue_state 
----+------------+-------------
(0 rows)

ROLLBACK;
